std: refactor the TLS implementation

joboet opened the pull request:
As discovered by Mara in #110897, our TLS implementation is a total mess. In the past months, I have simplified the actual macros and their expansions, but the majority of the complexity comes from the platform-specific support code needed to create keys and register destructors. In keeping with #117276, I have therefore moved all of the `thread_local_key`/`thread_local_dtor` modules to the `thread_local` module in `sys` and merged them into a new structure, so that future porters of `std` can simply mix-and-match the existing code instead of having to copy the same (bad) implementation everywhere. The new structure should become obvious when looking at `sys/thread_local/mod.rs`.

Unfortunately, the documentation changes associated with the refactoring have made this PR rather large. That said, this contains no functional changes except for two small ones:
* the key-based destructor fallback now, by virtue of sharing the implementation used by macOS and others, stores its list in a `#[thread_local]` static instead of in the key, eliminating one indirection layer and drastically simplifying its code.
* I've switched over ZKVM (tier 3) to use the same implementation as WebAssembly, as the implementation was just a way worse version of that

Please let me know if I can make this easier to review! I know these large PRs aren't optimal, but I couldn't think of any good intermediate steps.

@rustbot label +A-thread-locals

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9529177239/job/26267633582) [(plain)](https://github.com/rust-lang/rust/commit/ff3e202257bf8bdfdfce960d9ae9bc16244022f7/checks/26267633582/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

joboet committed:
std: refactor the TLS implementation

As discovered by Mara in #110897, our TLS implementation is a total mess. In the past months, I have simplified the actual macros and their expansions, but the majority of the complexity comes from the platform-specific support code needed to create keys and register destructors. In keeping with #117276, I have therefore moved all of the `thread_local_key`/`thread_local_dtor` modules to the `thread_local` module in `sys` and merged them into a new structure, so that future porters of `std` can simply mix-and-match the existing code instead of having to copy the same (bad) implementation everywhere. The new structure should become obvious when looking at `sys/thread_local/mod.rs`.

Unfortunately, the documentation changes associated with the refactoring have made this PR rather large. That said, this contains no functional changes except for two small ones:
* the key-based destructor fallback now, by virtue of sharing the implementation used by macOS and others, stores its list in a `#[thread_local]` static instead of in the key, eliminating one indirection layer and drastically simplifying its code.
* I've switched over ZKVM (tier 3) to use the same implementation as WebAssembly, as the implementation was just a way worse version of that

Please let me know if I can make this easier to review! I know these large PRs aren't optimal, but I couldn't think of any good intermediate steps.

@rustbot label +A-thread-locals

Mark-Simulacrum reviewed (changes requested):
I didn't review the implementations in detail (I think they are just moved, not actually changed, right?) -- left some comments inline, but most are non-blocking, just curiosity/slight improvements.

In general I'm happy to r+ with some of the nits fixed/questions answered.

ChrisDenton commented on library/std/src/sys/pal/windows/c.rs:
```diff
@@ -54,6 +54,7 @@ pub const EXIT_FAILURE: u32 = 1;
 pub const CONDITION_VARIABLE_INIT: CONDITION_VARIABLE = CONDITION_VARIABLE { Ptr: ptr::null_mut() };
 #[cfg(target_vendor = "win7")]
 pub const SRWLOCK_INIT: SRWLOCK = SRWLOCK { Ptr: ptr::null_mut() };
+#[cfg(not(target_thread_local))]
```
It's very intentional. `windows-gnu` uses an emulated TLS rather than native TLS. We've tried using native TLS regardless but everything breaks.

joboet committed:
std: simplify `#[cfg]`s for TLS

joboet commented on library/std/src/sys/thread_local/mod.rs:
```diff
@@ -1,27 +1,135 @@
+            pub(super) use linux::register;
+            pub(super) use list::run;
+        } else if #[cfg(all(
+            target_thread_local,
```
Sure, that makes sense.

joboet commented on library/std/src/sys/thread_local/mod.rs:
```diff
@@ -1,27 +1,135 @@
+            mod list;
+            pub(super) use list::register;
+            pub(crate) use list::run;
+        }
```
If a target has native thread local support, it needs a way to register destructors.

joboet commented on library/std/src/sys/thread_local/mod.rs:
```diff
@@ -1,27 +1,135 @@
+                target_os = "dragonfly"
+            )
+        ))] {
+            mod linux;
```
I wasn't sure about naming here, since this is only a subset of the UNIX platforms. How about `linux_like` instead?

joboet commented on library/std/src/sys/thread_local/destructors/linux.rs:
```diff
@@ -0,0 +1,58 @@
+    #[cfi_encoding = "i"]
+    #[repr(transparent)]
+    #[allow(non_camel_case_types)]
+    pub struct c_int(#[allow(dead_code)] pub libc::c_int);
```
I don't think so.

joboet committed:
std: use the `c_int` from `core::ffi` instead of `libc`

joboet commented on library/std/src/sys/thread_local/key/racy.rs:
```diff
@@ -126,18 +76,18 @@ impl StaticKey {
 
     #[inline]
-    unsafe fn key(&self) -> imp::Key {
+    fn key(&self) -> super::Key {
```
Nothing – it shouldn't have been `unsafe` to begin with, as key creation doesn't have any preconditions to uphold. Since I enabled `#[forbid(unsafe_op_in_unsafe_fn)]`, I decided to remove the `unsafe` here instead of adding a bunch of unnecessary ones elsewhere.

joboet commented on library/std/src/sys/thread_local/key/racy.rs:
```diff
@@ -126,18 +76,18 @@ impl StaticKey {
     }
 
-    unsafe fn lazy_init(&self) -> usize {
+    fn lazy_init(&self) -> usize {
```
Same thing – it's always safe.

joboet committed:
std: update TLS module documentation

joboet committed:
std: rename module for clarity

joboet commented:
I updated the documentation, it should be a bit clearer now.
@rustbot ready

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 32f9b8bf7671407fc19d9635b7c71b8051b8c165 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 32f9b8bf7671407fc19d9635b7c71b8051b8c165 -->
<!-- homu: {"type":"Approved","sha":"32f9b8bf7671407fc19d9635b7c71b8051b8c165","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

compiler-errors commented:
@bors r-

Please read the logs for https://github.com/rust-lang/rust/pull/126867#issuecomment-2185160546

Namely:

```
2024-06-23T17:09:09.6627017Z [0m[1m[38;5;9merror[0m[0m[1m: unused import: `list::run`[0m
2024-06-23T17:09:09.6628356Z [0m  [0m[0m[1m[38;5;12m--> [0m[0mlibrary/std/src/sys/thread_local/mod.rs:69:28[0m
2024-06-23T17:09:09.6629355Z [0m   [0m[0m[1m[38;5;12m|[0m
2024-06-23T17:09:09.6630612Z [0m[1m[38;5;12m69[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0m            pub(crate) use list::run;[0m
2024-06-23T17:09:09.6632017Z [0m   [0m[0m[1m[38;5;12m|[0m[0m                            [0m[0m[1m[38;5;9m^^^^^^^^^[0m
2024-06-23T17:09:09.6632995Z [0m   [0m[0m[1m[38;5;12m|[0m
2024-06-23T17:09:09.6634110Z [0m   [0m[0m[1m[38;5;12m= [0m[0m[1mnote[0m[0m: `-D unused-imports` implied by `-D warnings`[0m
2024-06-23T17:09:09.6635703Z [0m   [0m[0m[1m[38;5;12m= [0m[0m[1mhelp[0m[0m: to override `-D warnings` add `#[allow(unused_imports)]`[0m
2024-06-23T17:09:09.6636596Z 
2024-06-23T17:09:11.6895910Z [0m[1m[38;5;9merror[0m[0m[1m: function `run` is never used[0m
2024-06-23T17:09:11.6897039Z [0m  [0m[0m[1m[38;5;12m--> [0m[0mlibrary/std/src/sys/thread_local/destructors/list.rs:27:15[0m
2024-06-23T17:09:11.6897838Z [0m   [0m[0m[1m[38;5;12m|[0m
2024-06-23T17:09:11.6898571Z [0m[1m[38;5;12m27[0m[0m [0m[0m[1m[38;5;12m|[0m[0m [0m[0mpub unsafe fn run() {[0m
2024-06-23T17:09:11.6899460Z [0m   [0m[0m[1m[38;5;12m|[0m[0m               [0m[0m[1m[38;5;9m^^^[0m
2024-06-23T17:09:11.6900088Z [0m   [0m[0m[1m[38;5;12m|[0m
2024-06-23T17:09:11.6900860Z [0m   [0m[0m[1m[38;5;12m= [0m[0m[1mnote[0m[0m: `-D dead-code` implied by `-D warnings`[0m
2024-06-23T17:09:11.6901949Z [0m   [0m[0m[1m[38;5;12m= [0m[0m[1mhelp[0m[0m: to override `-D warnings` add `#[allow(dead_code)]`[0m
2024-06-23T17:09:11.6902565Z 
2024-06-23T17:09:11.8121882Z [RUSTC-TIMING] std test:false 2.411
```

joboet committed:
std: fix wasm builds

joboet commented:
Ah, I forgot to test that target. It should be fixed now, and I was able to carry over the note about wasm from `sys/unsupported/thread_local_dtor.rs`.

@rustbot ready

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit a2f078dbb55713068d7f61851f8e90a1a0556c05 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum a2f078dbb55713068d7f61851f8e90a1a0556c05 -->
<!-- homu: {"type":"Approved","sha":"a2f078dbb55713068d7f61851f8e90a1a0556c05","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

workingjubilee commented:
@bors rollup=never

Big refactor of TLS, don't want to have to guess when bisecting to this.

bors commented:
:hourglass: Testing commit a2f078dbb55713068d7f61851f8e90a1a0556c05 with merge 45edc40db940fb05f7eedf46fcae73ff4f8b1a54...
<!-- homu: {"type":"BuildStarted","head_sha":"a2f078dbb55713068d7f61851f8e90a1a0556c05","merge_sha":"45edc40db940fb05f7eedf46fcae73ff4f8b1a54"} -->

rust-log-analyzer commented:
The job **`dist-various-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9639491171/job/26581758433) [(plain)](https://github.com/rust-lang-ci/rust/commit/45edc40db940fb05f7eedf46fcae73ff4f8b1a54/checks/26581758433/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

error: unused import: `racy::StaticKey`
   --> library/std/src/sys/thread_local/mod.rs:125:28
    |
125 |             pub(super) use racy::StaticKey;
    |
    = note: `-D unused-imports` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(unused_imports)]`

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9639491171/job/26582700599)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9639491171/job/26582700599","builder_name":"checks-actions"} -->

joboet commented:
I forgot that Emscripten is also in the UNIX family...
@bors r=@Mark-Simulacrum

bors commented:
:pushpin: Commit 50a02ed789ed164f09d3ec8471f71c9c2b02df1d has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 50a02ed789ed164f09d3ec8471f71c9c2b02df1d -->
<!-- homu: {"type":"Approved","sha":"50a02ed789ed164f09d3ec8471f71c9c2b02df1d","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 50a02ed789ed164f09d3ec8471f71c9c2b02df1d with merge 5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4...
<!-- homu: {"type":"BuildStarted","head_sha":"50a02ed789ed164f09d3ec8471f71c9c2b02df1d","merge_sha":"5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9648497963/job/26616581105)
Approved by: Mark-Simulacrum
Pushing 5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9648497963/job/26616581105"},"merge_sha":"5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4"} -->

rust-timer commented:
Finished benchmarking commit (5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4): [comparison URL](https://perf.rust-lang.org/compare.html?start=d371d17496f2ce3a56da76aa083f4ef157572c20&end=5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d371d17496f2ce3a56da76aa083f4ef157572c20&end=5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d371d17496f2ce3a56da76aa083f4ef157572c20&end=5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=d371d17496f2ce3a56da76aa083f4ef157572c20&end=5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=d371d17496f2ce3a56da76aa083f4ef157572c20&end=5a3e2a4e921097c8f2bf6ea7565f8abe878cdbd4&stat=size:linked_artifact)
<details>
<summary>Results (primary -0.1%, secondary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.1%, -0.1%] | 4     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.0%] | 35    |
| All ❌✅ (primary)                 | -0.1% | [-0.1%, -0.1%] | 4     |
</details>

**Bootstrap**: 692.546s -> 692.589s (0.01%)
**Artifact size**: 326.77 MiB -> 326.77 MiB (-0.00%)
